File: tasks/c-relay-hop-audit/solution/relay_audit.c
#include <cjson/cJSON.h>
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_HOPS 32
#define MAX_FLOWS 64
#define MAX_EPOCHS 16
#define MAX_INCIDENTS 64
#define MAX_ADM 128
#define MAX_DEN 128
#define MAX_LED 128

typedef struct {
  char flow_id[64];
  int epoch;
  char hop_id[32];
  int bytes;
} Flow;

typedef struct {
  char kind[32];
  int epoch;
  char hop_id[32];
  int delta;
} Incident;

static char *read_file(const char *path) {
  FILE *f = fopen(path, "rb");
  if (!f) return NULL;
  if (fseek(f, 0, SEEK_END) != 0) {
    fclose(f);
    return NULL;
  }
  long sz = ftell(f);
  if (sz < 0) {
    fclose(f);
    return NULL;
  }
  rewind(f);
  char *buf = malloc((size_t)sz + 1);
  if (!buf) {
    fclose(f);
    return NULL;
  }
  if (fread(buf, 1, (size_t)sz, f) != (size_t)sz) {
    free(buf);
    fclose(f);
    return NULL;
  }
  buf[sz] = '\0';
  fclose(f);
  return buf;
}

static cJSON *load_json(const char *path) {
  char *raw = read_file(path);
  if (!raw) return NULL;
  cJSON *j = cJSON_Parse(raw);
  free(raw);
  return j;
}

static int env_or(const char *key, const char *defv, char *out, size_t outsz) {
  const char *v = getenv(key);
  const char *use = v && v[0] ? v : defv;
  snprintf(out, outsz, "%s", use);
  return 0;
}

static int cmp_path_row(const void *a, const void *b) {
  return strcmp((const char *)a, (const char *)b);
}

static int list_json_paths(const char *dir, char paths[][512], int maxn) {
  DIR *d = opendir(dir);
  if (!d) return 0;
  int n = 0;
  struct dirent *ent;
  while ((ent = readdir(d)) != NULL && n < maxn) {
    size_t len = strlen(ent->d_name);
    if (len < 6 || strcmp(ent->d_name + len - 5, ".json") != 0) continue;
    snprintf(paths[n], 512, "%s/%s", dir, ent->d_name);
    n++;
  }
  closedir(d);
  qsort(paths, (size_t)n, sizeof(paths[0]), cmp_path_row);
  return n;
}

static int cap_core(const char *hop, int *base, int *delta, int *halted,
                    const char hops_order[][32], int nh) {
  int hidx = -1;
  for (int i = 0; i < nh; i++) {
    if (strcmp(hops_order[i], hop) == 0) {
      hidx = i;
      break;
    }
  }
  if (hidx < 0) return 0;
  if (halted[hidx]) return 0;
  int v = base[hidx] + delta[hidx];
  return v < 1 ? 1 : v;
}

static int imin(int a, int b) { return a < b ? a : b; }
static int imax(int a, int b) { return a > b ? a : b; }

int main(void) {
  char data_dir[512], audit_dir[512];
  env_or("CRA_DATA_DIR", "/app/relayhop", data_dir, sizeof data_dir);
  env_or("CRA_AUDIT_DIR", "/app/audit", audit_dir, sizeof audit_dir);

  char pol_path[768];
  snprintf(pol_path, sizeof pol_path, "%s/policy.json", data_dir);
  cJSON *pol = load_json(pol_path);
  if (!pol) return 1;

  int carry_max = cJSON_GetObjectItem(pol, "carry_max")->valueint;
  cJSON *epochs_j = cJSON_GetObjectItem(pol, "epochs");
  cJSON *hops_j = cJSON_GetObjectItem(pol, "hops_order");
  int nepochs = cJSON_GetArraySize(epochs_j);
  int nh = cJSON_GetArraySize(hops_j);
  if (nh > MAX_HOPS || nepochs > MAX_EPOCHS) return 1;

  int epochs[MAX_EPOCHS];
  char hops_order[MAX_HOPS][32];
  for (int i = 0; i < nepochs; i++)
    epochs[i] = cJSON_GetArrayItem(epochs_j, i)->valueint;
  for (int i = 0; i < nh; i++)
    snprintf(hops_order[i], 32, "%s",
             cJSON_GetArrayItem(hops_j, i)->valuestring);

  char inc_path[768];
  snprintf(inc_path, sizeof inc_path, "%s/incidents.json", data_dir);
  cJSON *inc_file = load_json(inc_path);
  if (!inc_file) return 1;
  cJSON *incidents = cJSON_GetObjectItem(inc_file, "incidents");
  int ninc = cJSON_GetArraySize(incidents);

  int base[MAX_HOPS] = {0};
  char hop_paths[32][512];
  char hops_dir[768];
  snprintf(hops_dir, sizeof hops_dir, "%s/hops", data_dir);
  int nhfiles = list_json_paths(hops_dir, hop_paths, 32);
  for (int i = 0; i < nhfiles; i++) {
    cJSON *hf = load_json(hop_paths[i]);
    if (!hf) return 1;
    const char *hid = cJSON_GetObjectItem(hf, "hop_id")->valuestring;
    int bc = cJSON_GetObjectItem(hf, "base_cap")->valueint;
    for (int j = 0; j < nh; j++) {
      if (strcmp(hops_order[j], hid) == 0) {
        base[j] = bc;
        break;
      }
    }
    cJSON_Delete(hf);
  }

  Flow flows[MAX_FLOWS];
  int nflows = 0;
  char flow_paths[64][512];
  char flows_dir[768];
  snprintf(flows_dir, sizeof flows_dir, "%s/flows", data_dir);
  int nffiles = list_json_paths(flows_dir, flow_paths, 64);
  for (int i = 0; i < nffiles && nflows < MAX_FLOWS; i++) {
    cJSON *ff = load_json(flow_paths[i]);
    if (!ff) return 1;
    snprintf(flows[nflows].flow_id, 64, "%s",
             cJSON_GetObjectItem(ff, "flow_id")->valuestring);
    flows[nflows].epoch = cJSON_GetObjectItem(ff, "epoch")->valueint;
    snprintf(flows[nflows].hop_id, 32, "%s",
             cJSON_GetObjectItem(ff, "hop_id")->valuestring);
    flows[nflows].bytes = cJSON_GetObjectItem(ff, "bytes")->valueint;
    nflows++;
    cJSON_Delete(ff);
  }

  int delta[MAX_HOPS] = {0};
  int halted[MAX_HOPS] = {0};
  int carry[MAX_HOPS] = {0};
  for (int i = 0; i < nh; i++) {
    delta[i] = 0;
    halted[i] = 0;
    carry[i] = 0;
  }

  cJSON *admissions = cJSON_CreateArray();
  cJSON *denials = cJSON_CreateArray();
  cJSON *ledgers = cJSON_CreateArray();

  for (int ei = 0; ei < nepochs; ei++) {
    int e = epochs[ei];
    for (int ii = 0; ii < ninc; ii++) {
      cJSON *inc = cJSON_GetArrayItem(incidents, ii);
      if (cJSON_GetObjectItem(inc, "epoch")->valueint != e) continue;
      const char *kind = cJSON_GetObjectItem(inc, "kind")->valuestring;
      if (strcmp(kind, "noop") == 0) continue;
      cJSON *hop_item = cJSON_GetObjectItem(inc, "hop_id");
      if (!hop_item || !cJSON_IsString(hop_item)) return 1;
      const char *hid = hop_item->valuestring;
      int hidx = -1;
      for (int j = 0; j < nh; j++) {
        if (strcmp(hops_order[j], hid) == 0) {
          hidx = j;
          break;
        }
      }
      if (strcmp(kind, "cap_add") == 0) {
        if (hidx >= 0)
          delta[hidx] += cJSON_GetObjectItem(inc, "delta")->valueint;
      } else if (strcmp(kind, "halt_hop") == 0) {
        if (hidx >= 0) {
          halted[hidx] = 1;
          carry[hidx] = 0;
        }
      } else if (strcmp(kind, "resume_hop") == 0) {
        if (hidx >= 0) {
          halted[hidx] = 0;
          carry[hidx] = 0;
        }
      } else {
        return 1;
      }
    }

    int cin[MAX_HOPS];
    int used[MAX_HOPS];
    for (int i = 0; i < nh; i++) {
      cin[i] = carry[i];
      used[i] = 0;
    }

    int epoch_flow_idx[MAX_FLOWS];
    int nepoch_flows = 0;
    for (int fi = 0; fi < nflows; fi++) {
      if (flows[fi].epoch == e) epoch_flow_idx[nepoch_flows++] = fi;
    }
    for (int a = 0; a < nepoch_flows; a++) {
      for (int b = a + 1; b < nepoch_flows; b++) {
        Flow *fa = &flows[epoch_flow_idx[a]];
        Flow *fb = &flows[epoch_flow_idx[b]];
        int swap = 0;
        if (strcmp(fa->hop_id, fb->hop_id) < 0)
          swap = 0;
        else if (strcmp(fa->hop_id, fb->hop_id) > 0)
          swap = 1;
        else if (strcmp(fa->flow_id, fb->flow_id) > 0)
          swap = 1;
        if (swap) {
          int t = epoch_flow_idx[a];
          epoch_flow_idx[a] = epoch_flow_idx[b];
          epoch_flow_idx[b] = t;
        }
      }
    }

    for (int fi = 0; fi < nepoch_flows; fi++) {
      Flow *f = &flows[epoch_flow_idx[fi]];
      int hidx = -1;
      for (int j = 0; j < nh; j++) {
        if (strcmp(hops_order[j], f->hop_id) == 0) {
          hidx = j;
          break;
        }
      }
      if (hidx < 0) continue;
      int cc = cap_core(f->hop_id, base, delta, halted, hops_order, nh);
      int bud = cc + cin[hidx];
      int avail = bud - used[hidx];
      if (avail < 0) avail = 0;
      if (f->bytes <= avail) {
        used[hidx] += f->bytes;
        cJSON *row = cJSON_CreateObject();
        cJSON_AddNumberToObject(row, "bytes", f->bytes);
        cJSON_AddNumberToObject(row, "epoch", e);
        cJSON_AddStringToObject(row, "flow_id", f->flow_id);
        cJSON_AddStringToObject(row, "hop_id", f->hop_id);
        cJSON_AddItemToArray(admissions, row);
      } else {
        cJSON *row = cJSON_CreateObject();
        cJSON_AddNumberToObject(row, "available", avail);
        cJSON_AddNumberToObject(row, "epoch", e);
        cJSON_AddStringToObject(row, "flow_id", f->flow_id);
        cJSON_AddStringToObject(row, "hop_id", f->hop_id);
        cJSON_AddNumberToObject(row, "requested", f->bytes);
        cJSON_AddItemToArray(denials, row);
      }
    }

    for (int hi = 0; hi < nh; hi++) {
      int cc = cap_core(hops_order[hi], base, delta, halted, hops_order, nh);
      int bud = cc + cin[hi];
      int u = used[hi];
      int rem = bud - u;
      int cout = imin(carry_max, imax(0, rem));
      if (halted[hi]) cout = 0;
      cJSON *row = cJSON_CreateObject();
      cJSON_AddNumberToObject(row, "cap_core", cc);
      cJSON_AddNumberToObject(row, "carry_in", cin[hi]);
      cJSON_AddNumberToObject(row, "carry_out", cout);
      cJSON_AddNumberToObject(row, "epoch", e);
      cJSON_AddStringToObject(row, "hop_id", hops_order[hi]);
      cJSON_AddNumberToObject(row, "used", u);
      cJSON_AddItemToArray(ledgers, row);
      carry[hi] = cout;
    }
  }

  cJSON *applied = cJSON_CreateArray();
  for (int ii = 0; ii < ninc; ii++) {
    cJSON *inc = cJSON_GetArrayItem(incidents, ii);
    cJSON_AddItemToArray(
        applied,
        cJSON_CreateString(cJSON_GetObjectItem(inc, "kind")->valuestring));
  }

  int max_ep = 0;
  for (int ii = 0; ii < ninc; ii++) {
    cJSON *inc = cJSON_GetArrayItem(incidents, ii);
    int ep = cJSON_GetObjectItem(inc, "epoch")->valueint;
    if (ep > max_ep) max_ep = ep;
  }
  cJSON *child;
  cJSON_ArrayForEach(child, admissions) {
    int ep = cJSON_GetObjectItem(child, "epoch")->valueint;
    if (ep > max_ep) max_ep = ep;
  }
  cJSON_ArrayForEach(child, denials) {
    int ep = cJSON_GetObjectItem(child, "epoch")->valueint;
    if (ep > max_ep) max_ep = ep;
  }

  int tot_adm = 0, tot_adm_bytes = 0;
  cJSON_ArrayForEach(child, admissions) {
    tot_adm++;
    tot_adm_bytes += cJSON_GetObjectItem(child, "bytes")->valueint;
  }
  int tot_den = cJSON_GetArraySize(denials);
  int tot_den_bytes = 0;
  cJSON_ArrayForEach(child, denials) {
    tot_den_bytes += cJSON_GetObjectItem(child, "requested")->valueint;
  }

  cJSON *summary = cJSON_CreateObject();
  cJSON_AddItemToObject(summary, "incidents_applied", applied);
  cJSON_AddNumberToObject(summary, "max_epoch", max_ep);
  cJSON_AddNumberToObject(summary, "total_admissions", tot_adm);
  cJSON_AddNumberToObject(summary, "total_admitted_bytes", tot_adm_bytes);
  cJSON_AddNumberToObject(summary, "total_denials", tot_den);
  cJSON_AddNumberToObject(summary, "total_denied_bytes", tot_den_bytes);

  char out_path[768];
  snprintf(out_path, sizeof out_path, "%s/admissions.json", audit_dir);
  cJSON *root = cJSON_CreateObject();
  cJSON_AddItemToObject(root, "admissions", cJSON_Duplicate(admissions, 1));
  char *printed = cJSON_Print(root);
  FILE *wf = fopen(out_path, "w");
  if (printed && wf) {
    fputs(printed, wf);
    fputc('\n', wf);
    fclose(wf);
  }
  free(printed);
  cJSON_Delete(root);

  snprintf(out_path, sizeof out_path, "%s/denials.json", audit_dir);
  root = cJSON_CreateObject();
  cJSON_AddItemToObject(root, "denials", cJSON_Duplicate(denials, 1));
  printed = cJSON_Print(root);
  wf = fopen(out_path, "w");
  if (printed && wf) {
    fputs(printed, wf);
    fputc('\n', wf);
    fclose(wf);
  }
  free(printed);
  cJSON_Delete(root);

  snprintf(out_path, sizeof out_path, "%s/carry_ledgers.json", audit_dir);
  root = cJSON_CreateObject();
  cJSON_AddItemToObject(root, "rows", cJSON_Duplicate(ledgers, 1));
  printed = cJSON_Print(root);
  wf = fopen(out_path, "w");
  if (printed && wf) {
    fputs(printed, wf);
    fputc('\n', wf);
    fclose(wf);
  }
  free(printed);
  cJSON_Delete(root);

  snprintf(out_path, sizeof out_path, "%s/summary.json", audit_dir);
  printed = cJSON_Print(summary);
  wf = fopen(out_path, "w");
  if (printed && wf) {
    fputs(printed, wf);
    fputc('\n', wf);
    fclose(wf);
  }
  free(printed);

  cJSON_Delete(pol);
  cJSON_Delete(inc_file);
  cJSON_Delete(admissions);
  cJSON_Delete(denials);
  cJSON_Delete(ledgers);
  cJSON_Delete(summary);
  return 0;
}

File: tools/batch15_src/relay_audit.py
#!/usr/bin/env python3
"""Reference relay-hop audit (used by py-relay-hop and hash-lock generation)."""

from __future__ import annotations

import json
import os
from pathlib import Path


def _env(key: str, default: str) -> str:
    v = os.environ.get(key, "").strip()
    return v if v else default


def _read_json(path: Path) -> object:
    return json.loads(path.read_text(encoding="utf-8"))


def _write_json(path: Path, obj: object) -> None:
    path.write_text(
        json.dumps(obj, ensure_ascii=True, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )


def _sorted_json_paths(directory: Path) -> list[Path]:
    if not directory.is_dir():
        return []
    return sorted(p for p in directory.glob("*.json") if p.is_file())


def _cap_core(hop: str, base: dict[str, int], delta: dict[str, int], halted: dict[str, bool]) -> int:
    if halted.get(hop, False):
        return 0
    v = base.get(hop, 0) + delta.get(hop, 0)
    return 1 if v < 1 else v


def main() -> None:
    data = Path(_env("RELAY_DATA_DIR", "/app/relayhop"))
    audit = Path(_env("RELAY_AUDIT_DIR", "/app/audit"))
    audit.mkdir(parents=True, exist_ok=True)

    pol = _read_json(data / "policy.json")
    assert isinstance(pol, dict)
    inc_file = _read_json(data / "incidents.json")
    assert isinstance(inc_file, dict)
    carry_max = int(pol["carry_max"])
    epochs = [int(x) for x in pol["epochs"]]
    hops_order = [str(x) for x in pol["hops_order"]]

    base: dict[str, int] = {}
    for p in _sorted_json_paths(data / "hops"):
        hf = _read_json(p)
        assert isinstance(hf, dict)
        base[str(hf["hop_id"])] = int(hf["base_cap"])

    flows: list[dict[str, object]] = []
    for p in _sorted_json_paths(data / "flows"):
        ff = _read_json(p)
        assert isinstance(ff, dict)
        flows.append(
            {
                "flow_id": str(ff["flow_id"]),
                "epoch": int(ff["epoch"]),
                "hop_id": str(ff["hop_id"]),
                "bytes": int(ff["bytes"]),
            }
        )

    epoch_set = set(epochs)
    hop_set = set(hops_order)
    for f in flows:
        if f["epoch"] not in epoch_set or f["hop_id"] not in base:
            raise SystemExit(1)
    if len(hop_set) != len(hops_order) or len(hop_set) != len(base):
        raise SystemExit(1)
    for h in base:
        if h not in hop_set:
            raise SystemExit(1)

    delta = {h: 0 for h in hops_order}
    carry = {h: 0 for h in hops_order}
    halted = {h: False for h in hops_order}

    admissions: list[dict[str, object]] = []
    denials: list[dict[str, object]] = []
    ledgers: list[dict[str, object]] = []

    incidents = inc_file.get("incidents", [])
    assert isinstance(incidents, list)

    for e in epochs:
        for inc in incidents:
            assert isinstance(inc, dict)
            if int(inc["epoch"]) != e:
                continue
            kind = str(inc["kind"])
            if kind == "noop":
                continue
            h = str(inc["hop_id"])
            if kind == "cap_add":
                delta[h] = delta.get(h, 0) + int(inc["delta"])
            elif kind == "halt_hop":
                halted[h] = True
                carry[h] = 0
            elif kind == "resume_hop":
                halted[h] = False
                carry[h] = 0
            else:
                raise SystemExit(1)

        cin = {h: carry[h] for h in hops_order}
        used = {h: 0 for h in hops_order}

        epoch_flows = [f for f in flows if f["epoch"] == e]
        epoch_flows.sort(key=lambda f: (str(f["hop_id"]), str(f["flow_id"])))

        for f in epoch_flows:
            h = str(f["hop_id"])
            b = int(f["bytes"])
            avail = _cap_core(h, base, delta, halted) + cin[h] - used[h]
            if avail < 0:
                avail = 0
            if b <= avail:
                used[h] += b
                admissions.append(
                    {
                        "bytes": b,
                        "epoch": e,
                        "flow_id": str(f["flow_id"]),
                        "hop_id": h,
                    }
                )
            else:
                denials.append(
                    {
                        "available": avail,
                        "epoch": e,
                        "flow_id": str(f["flow_id"]),
                        "hop_id": h,
                        "requested": b,
                    }
                )

        for h in hops_order:
            cc = _cap_core(h, base, delta, halted)
            rem = cc + cin[h] - used[h]
            cout = min(carry_max, max(0, rem))
            if halted[h]:
                cout = 0
            ledgers.append(
                {
                    "cap_core": cc,
                    "carry_in": cin[h],
                    "carry_out": cout,
                    "epoch": e,
                    "hop_id": h,
                    "used": used[h],
                }
            )
            carry[h] = cout

    admissions.sort(key=lambda r: (int(r["epoch"]), str(r["hop_id"]), str(r["flow_id"])))
    denials.sort(key=lambda r: (int(r["epoch"]), str(r["hop_id"]), str(r["flow_id"])))
    ledgers.sort(key=lambda r: (int(r["epoch"]), str(r["hop_id"])))

    applied = [str(inc["kind"]) for inc in incidents if isinstance(inc, dict)]
    max_ep = 0
    for inc in incidents:
        if isinstance(inc, dict):
            max_ep = max(max_ep, int(inc["epoch"]))
    for row in admissions + denials:
        max_ep = max(max_ep, int(row["epoch"]))

    tot_adm = len(admissions)
    tot_adm_bytes = sum(int(r["bytes"]) for r in admissions)
    tot_den = len(denials)
    tot_den_bytes = sum(int(r["requested"]) for r in denials)

    _write_json(audit / "admissions.json", {"admissions": admissions})
    _write_json(audit / "denials.json", {"denials": denials})
    _write_json(audit / "carry_ledgers.json", {"rows": ledgers})
    _write_json(
        audit / "summary.json",
        {
            "incidents_applied": applied,
            "max_epoch": max_ep,
            "total_admissions": tot_adm,
            "total_admitted_bytes": tot_adm_bytes,
            "total_denials": tot_den,
            "total_denied_bytes": tot_den_bytes,
        },
    )


if __name__ == "__main__":
    main()
